vms/proposervm/block/build.go

```diff
@@ -6,75 +6,169 @@ package block
 	timestamp time.Time,
 	pChainHeight uint64,
 	blockBytes []byte,
+	parentBlockSig []byte,
 ) (SignedBlock, error) {
-	var block SignedBlock = &statelessBlock{
+	// in this case, we can't sign with BLS key, since we're not going to include the Certificate, which is required
+	// for the signature validation. Instead, we'll just hash the previous
+	sigParentBlockSig := nextHashBlockSignature(parentBlockSig)
+
+	block := &statelessBlock{
 		StatelessBlock: statelessUnsignedBlock{
 			ParentID:     parentID,
 			Timestamp:    timestamp.Unix(),
 			PChainHeight: pChainHeight,
 			Certificate:  nil,
 			Block:        blockBytes,
+			VRFSig:       sigParentBlockSig,
 		},
 		timestamp: timestamp,
 	}
-
-	bytes, err := Codec.Marshal(CodecVersion, &block)
+	bytes, err := marshalBlock(block)
 	if err != nil {
 		return nil, err
 	}
 
 	return block, block.initialize(bytes)
 }
 
+// marshalBlock marshal the given statelessBlock by using either the default statelessBlock or
+// coping the exported fields into statelessBlockV0 and then marshaling it.
+// this allows the marsheler to produce encoded blocks that match the old style blocks as long as
+// the VRGSig feature was not enabled.
+func marshalBlock(block *statelessBlock) ([]byte, error) {
+	if len(block.StatelessBlock.VRFSig) == 0 {
+		// create a backward compatible block ( without VRFSig ) and use the statelessBlockV0 encoder for the encoding.
+		var preBlockSigBlock SignedBlock = &statelessBlockV0{
+			StatelessBlock: statelessUnsignedBlockV0{
+				ParentID:     block.StatelessBlock.ParentID,
+				Timestamp:    block.StatelessBlock.Timestamp,
+				PChainHeight: block.StatelessBlock.PChainHeight,
+				Certificate:  block.StatelessBlock.Certificate,
+				Block:        block.StatelessBlock.Block,
+			},
+			Signature: block.Signature,
+		}
+		return Codec.Marshal(CodecVersion, &preBlockSigBlock)
+	}
+	var blockIntf SignedBlock = block
+	return Codec.Marshal(CodecVersion, &blockIntf)
+}
+
+func calculateBootstrappingBlockSig(chainID ids.ID, networkID uint32) [hashing.HashLen]byte {
+	// build the hash of the following struct:
+	// +-----------------------+----------+------------+
+	// |  prefix :             | [8]byte  | "rng-root" |
+	// +-----------------------+----------+------------+
+	// |  chainID :            | [32]byte |  32 bytes  |
+	// +-----------------------+----------+------------+
+	// |  networkID:           | uint32   |  4 bytes   |
+	// +-----------------------+----------+------------+
+
+	buffer := make([]byte, 44)
+	copy(buffer, "rng-root")
+	copy(buffer[8:], chainID[:])
+	binary.LittleEndian.PutUint32(buffer[40:], networkID)
+	return hashing.Hash256(buffer)
+}
+
+func nextHashBlockSignature(parentBlockSig []byte) []byte {
+	if len(parentBlockSig) == 0 {
+		return nil
+	}
+	// previous block had a valid signature, hash that signature.
+	sigParentBlockSig := hashing.ComputeHash256(parentBlockSig)
+
+	// as long as the signature length is too short, generate additional hashes.
```

Comment:
I'm a bit confused as to why we even need to pad this at all. Wouldn't it make more sense to just have it be the 32-byte slice? We already encode the length of the signature in the block, I'm not sure why we care to ensure it is always the same length.

Reply:
I was thinking about that too. The question that I tried to answer is whether an observer should have the ability to distinguish between a hash pseudo signature block vs a genuine signature.
Just padding the 32 bytes so it would fit into a 96 byte slice, doesn't give us any advantage.

Reply:
An observer will presumably have [access to all other relevant information](https://en.wikipedia.org/wiki/Kerckhoffs%27s_principle), including the fact that the proposer doesn't have a BLS key.